BE-8005: before and after

ajanata opened the pull request:
some other minor cleanup

added lots of tests, fixed a couple bugs the tests found

Commit:
BE-8005: before and after

some other minor cleanup

ajanata commented on orderedmap/map_test.go:
```diff
@@ -1,10 +1,10 @@
 )
 
-func TestOrderedMap(t *testing.T) {
+func ExampleOrderedMap() {
```
Examples are run as tests and will fail if the output does not match the provided output in the comments.

ajanata commented on register.go:
```diff
@@ -32,25 +23,158 @@ func RegisterTest(name string, tester Tester) any {
+// The return value may be discarded (and is always nil); it is provided to simplify writing test code, like so:
+//
+//    var _ = testy.Test("my test", func(t testy.TestingT){})
+func Test(name string, tester Tester) any {
```
Revisiting the name on this. Propose renaming this to align with `BeforePackage` etc.; `RegisterBeforePackage` is way too wordy.

ajanata commented on register.go:
```diff
@@ -32,25 +23,158 @@ func RegisterTest(name string, tester Tester) any {
 	return nil
 }
+
+// BeforePackage registers a function to be run once before any tests in the given package are run. A package may only
+// have one BeforePackage function.
```
This (and the others) could be expanded to allow multiple helpers, if that proves useful, but an execution order would have to be defined (likely the order that initialization occurs). That also can be done without breaking existing users, so I see no reason to worry about it until it does actually show itself to be useful.

kendallm replied:
How does this work out if you are testing across flies in the same package.

eg

file1.go
```
package gametime
```

file2.go
```
package gametime
```

I can see use cases where this is desirable and from reading the implementation it does not seem like it would work.

ajanata replied:
If they absolutely need that behavior, they can be in different packages. There isn't an easy way to get more fine grain control in a generic manner. Alternatively (and this is what we do a lot already), you can have common setup in the outer level test and have subtests that do the actual testing. These are designed more for stuff like spinning up containers (which we will need eventually) or creating entries in a database.

Commit:
constraints is getting moved to experimental

Commit:
bug fix, add tests

* getCallerPackage didn't work right for anonymous functions
* add tests for most things

Commit:
run tests on github

Commit:
version is not in standard go format...

Commit:
need stable: false too

Commit:
go fmt for 1.18

ajanata commented on orderedmap/map.go:
```diff
@@ -1,8 +1,9 @@
-	"constraints"
 	"sort"
+
+	"golang.org/x/exp/constraints"
```
`constraints` is going to be removed from the standard library in the next beta/rc build, but it has already been pushed to experimental, so making this change now

ajanata commented on run.go:
```diff
@@ -13,48 +15,223 @@ import (
-				tt.Helper()
-				test.tester(tWrapper{t: tt})
+	instance.tests.Iterate(func(pkg string, pkgTests *testPkg) bool {
+		// we have to hold onto any panics here to be able to run AfterPackage
```
this whole thing is kinda ugly and I can't think of a way to test it since it is supposed to run inside a test anyway.

ajanata commented on run.go:
```diff
@@ -13,48 +15,223 @@ import (
-				for r := range res {
-					results[pkg][r.Name] = r
-				}
+		// we have to hold onto any panics here to be able to run AfterPackage
```
and this is unfortunately even uglier, though it is rather extensively tested

Commit:
remove cruft from when RunFromTest was going to be tested

jeffatgametime commented on run.go:
```diff
@@ -13,48 +15,223 @@ import (
+					Passed:  false,
+					Dur:     result.Dur,
+					Msgs: append(result.Msgs, append(pkgHelperT.msgs, Msg{
+						Msg:   fmt.Sprintf("%v", beforePkgErr),
```
```suggestion
						Msg:   fmt.Sprintf("%v", afterPkgErr),
```

but why are we not reporting `afterPkgErr` if `beforePkgErr` is not nil?

ajanata replied:
can't panic two things, and if before failed any failures in after could very well be related.

ajanata replied:
wait, this isn't the path that panics, it should have already been reported above... though do the tests reflect that? hm.

ajanata replied:
ah, I don't have any test cases with multiple things failing, let me add one. but I think it should have them both

ajanata replied:
ok, yeah, this was easy enough to do in Run mode

Commit:
Update run.go

Co-authored-by: jeffatgametime <48690769+jeffatgametime@users.noreply.github.com>

kendallm commented on run.go:
```diff
@@ -13,48 +15,223 @@ import (
+		// we have to hold onto any panics here to be able to run AfterPackage
+		var beforePkgErr any
+		if pkgTests.BeforePackage != nil {
+			func() {
```
Is it necessary to wrap this in an anonymous function here?

ajanata replied:
yes, because the `defer` needs to apply only to `pkgTests.BeforePackage`

Commit:
report AfterPackage panics as well

and document to not call the failure funcs on the TestingT in
before/after funcs for now

kendallm reviewed (commented):
Still wrapping my head around the concurrency stuff, but LGTM based on the testing you have in here. Very helpful!

Couple questions though.
